Adin2111 add support

tachicialex opened the pull request:
The ADIN2111 is a low power, low complexity, two-Ethernet ports
switch with integrated 10BASE-T1L PHYs and one serial peripheral
interface (SPI) port.

ADIN2111 contains two ADIN1100 PHYs that are exposed to the user via an
internal MDIO bus. The bus can be accessed via read/write of the MAC's SPI
MDIO control registers.

The driver will register and allocate two struct net_dev structures called "port netdevs".
These are the highest level abstraction of a MAC's physical port.
In this way 

Using standard netdev tools (iproute2, ethtool, etc), the port netdev can also provide to
the user access to the physical properties of the switch port such as PHY link state and I/O statistics.

Ex:

```
adin2111-0-p0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 192.168.160.160  netmask 255.255.255.0  broadcast 192.168.160.255
        inet6 fe80::c82f:b7ff:fe10:2363  prefixlen 64  scopeid 0x20<link>
        ether ca:2f:b7:10:23:63  txqueuelen 1000  (Ethernet)
        RX packets 29  bytes 2400 (2.3 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 99  bytes 11666 (11.3 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device interrupt 199
adin2111-0-p1: flags=4419<UP,BROADCAST,RUNNING,PROMISC,MULTICAST>  mtu 1500
        inet 192.168.160.160  netmask 255.255.255.0  broadcast 192.168.160.255
        inet6 fe80::c82f:b7ff:fe10:2364  prefixlen 64  scopeid 0x20<link>
        ether ca:2f:b7:10:23:64  txqueuelen 1000  (Ethernet)
        RX packets 69  bytes 6696 (6.5 KiB)
        RX errors 2  dropped 0  overruns 0  frame 0
        TX packets 126  bytes 17366 (16.9 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device interrupt 199

```

nunojsa commented on Documentation/devicetree/bindings/net/adi,adin2111.yaml:
```diff
@@ -0,0 +1,120 @@
+                                reg = <1>;
+                        };
+                };
+        };
```
I think you might be able to add this to the adin1110.yaml file with some conditional. It's almost identical. On top of this, the device is also supported in the adin1110 driver so it really makes sense for it to be there

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -853,113 +1025,172 @@ static int adin1110_check_spi(struct adin1110_priv *priv)
-		return PTR_ERR(priv->phydev);
-
-	return devm_add_action_or_reset(dev, adin1110_disconnect_phy, priv->phydev);
+	return devm_add_action_or_reset(dev, adin1110_disconnect_phy, priv);
```
This is nok. This should be called right after each `phy_connect()`. You do a `devm_request_irq ()` after that which means that your unbinding path won't be the reverse order of your binding path. That is asking for problems

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -853,113 +1025,172 @@ static int adin1110_check_spi(struct adin1110_priv *priv)
+	ret = devm_request_threaded_irq(dev, priv->spidev->irq, NULL, adin1110_irq,
+					IRQF_TRIGGER_LOW | IRQF_ONESHOT, dev_name(dev), priv);
+	if (ret < 0)
+		dev_err(dev, "Failed to get IRQ: %d\n", ret);
```
don't see the point of this log... just `return devm_request_threaded_irq(...)`

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -853,113 +1025,172 @@ static int adin1110_check_spi(struct adin1110_priv *priv)
+		port_priv->phydev = phy_connect(netdev, phydev_name(port_priv->phydev),
+						adin1110_adjust_link, PHY_INTERFACE_MODE_MII);
+		if (IS_ERR(port_priv->phydev))
+			return PTR_ERR(port_priv->phydev);
```
as said, action to disconnect each phy should go here

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -853,113 +1025,172 @@ static int adin1110_check_spi(struct adin1110_priv *priv)
+	int i;
+
+	for (i = 0; i < priv->cfg->ports_nr; i++)
+		phy_disconnect(priv->ports[i]->phydev);
```
Another potential thing with your approach. You can successfully enable the first phy but not the second but you would still try to disable them both... Just use the original function you add to disconnect the phy and add 2 actions

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -763,53 +906,80 @@ static netdev_tx_t adin1110_start_xmit(struct sk_buff *skb, struct net_device *d
+	int err;
+
+	err = snprintf(name, len, "p%d", port_priv->nr);
+
```
nit: no new line

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -763,53 +906,80 @@ static netdev_tx_t adin1110_start_xmit(struct sk_buff *skb, struct net_device *d
 	storage->rx_dropped = val;
 
-	adin1110_read_reg(priv, ADIN1110_RX_MCAST_CNT, &val);
+	adin1110_read_reg(priv, ADIN1110_RX_MCAST_CNT + addr_offset, &val);
```
is it possible to detect that we need the offset inside `adin1110_read_reg()`? Might be neater

nunojsa reviewed (commented):
Looks good overall. Some stuff that needs to be addressed though

tachicialex commented:
Changelog v1 -> v2:
  - added conditional DT properties. Removed standalone adin2111.yaml. Moved everything in
  adin1110.yaml
  - in adin1110_probe_netdevs(): added separate devm_actions to disconect each phy only after a
  successful phy_connect()
  - removed logging of devm_request_threaded_irq()
  - added a small fix for TX_SPACE read (reporting a twice as low FIFO size was affecting bandwidth)
  - added macros in adin1110_ndo_get_stats64() instead of adding the offset on the spot

nunojsa commented on Documentation/devicetree/bindings/net/adi,adin1110.yaml:
```diff
@@ -1,7 +1,7 @@
 %YAML 1.2
 ---
-$id: "http://devicetree.org/schemas/net/adi,adin1110.yaml#"
+$id: "http://devicetree.org/schemas/bindings/net/adi,adin1110.yaml#"
```
I think that you are actually changing this to the wrong format... Please make sure you validate your binding

tachicialex commented:
Changelog v2 -> v3:
- rebased
- fixed checkpatch complaint about fix commit hash

nunojsa commented on Documentation/devicetree/bindings/net/adi,adin1110.yaml:
```diff
@@ -18,12 +18,23 @@ description: |
+      - items:
+          - enum:
+              - adi,adin1110
+              - adi,adin2111
```
This can simply be:

```
properties:
  compatible:
    enum:
      - adi,adin1110
      - adi,adin2111
```

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -854,113 +1024,168 @@ static int adin1110_check_spi(struct adin1110_priv *priv)
-		return PTR_ERR(priv->phydev);
-
-	return devm_add_action_or_reset(dev, adin1110_disconnect_phy, priv->phydev);
+	return 0;
```
nit: return `adin1110_probe_netdevs()`... Do all the meaningful logging inside the function

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -854,113 +1024,168 @@ static int adin1110_check_spi(struct adin1110_priv *priv)
-	priv = netdev_priv(netdev);
+	dev_id = spi_get_device_id(spi);
+	if (!dev_id)
+		return -EINVAL;
```
I don't think this can ever happen...

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -854,113 +1024,168 @@ static int adin1110_check_spi(struct adin1110_priv *priv)
+		if (IS_ERR(port_priv->phydev))
+			return PTR_ERR(port_priv->phydev);
+
+		devm_add_action_or_reset(dev, adin1110_disconnect_phy, port_priv);
```
error handling....

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -854,113 +1024,168 @@ static int adin1110_check_spi(struct adin1110_priv *priv)
+	int i;
+
+	for (i = 0; i < priv->cfg->ports_nr; i++) {
+		netdev = devm_alloc_etherdev(dev, sizeof(struct adin1110_port_priv));
```
nit: `sizeof(*port_priv)`

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -854,113 +1024,168 @@ static int adin1110_check_spi(struct adin1110_priv *priv)
-	phy_disconnect(data);
+	struct adin1110_port_priv *port_priv = data;
+
+	phy_disconnect(port_priv->phydev);
```
not that big deal but you can directly pass `port_priv->phydev` to this callback... Then you can keep the original from which is simpler

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -688,60 +828,62 @@ static int adin1110_net_open(struct net_device *net_dev)
 	int ret;
 
+	port_priv = container_of(work, struct adin1110_port_priv, tx_work);
+	priv = port_priv->priv;
```
nit: you could still do this when declaring the variables

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -688,60 +828,62 @@ static int adin1110_net_open(struct net_device *net_dev)
-	if (ret < 0)
+	ret = adin1110_init_mac(port_priv);
+	if (ret < 0) {
+		mutex_unlock(&priv->lock);
```
Hmmm... This is actually fixing a bug and is not related with this patch.

Please add the fix as a separate patch with the proper tag. Hiding fixes like this in unrelated commits is not a good practice

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -607,46 +719,64 @@ static int adin1110_ioctl(struct net_device *netdev, struct ifreq *rq, int cmd)
+	u32 mask;
+
+	port_priv = container_of(work, struct adin1110_port_priv, rx_mode_work);
+	priv = port_priv->priv;
```
nit: ditto

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -607,46 +719,64 @@ static int adin1110_ioctl(struct net_device *netdev, struct ifreq *rq, int cmd)
+		break;
+	default:
+		return;
+	}
```
any chance for `port_priv->nr` be not 0 or 1? Isn't this something that we can control at probe or maybe already enforced by static data? If so, this could simple be:

```
if (!port_priv->nr)
     mask = ADIN1110_FWD_UNK2HOST;
else
     mask = ADIN2111_P2_FWD_UNK2HOST;
```

Or you could even use the ternary operator even though I'm not a big fan (but on a simple assignment like this is more than fine).

(Just saying this because the default: statement is useless if we can guarantee what I've said)

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -236,18 +293,31 @@ static int adin1110_round_len(int len)
+		ret = adin1110_read_reg(priv, ADIN2111_RX_P2_FSIZE, &frame_size);
+		break;
+	default:
+		return -EINVAL;
```
ditto...

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -236,18 +293,27 @@ static int adin1110_round_len(int len)
+	if (!port_priv->nr) {
+		__reg = cpu_to_le16(ADIN1110_RX);
+		ret = adin1110_read_reg(priv, ADIN1110_RX_FSIZE, &frame_size);
+
```
nit: blank line not needed...

nunojsa reviewed (approved):
Just a minor nit...

tachicialex commented:
Changelog v3 -> v4:
- fixed unnecessary blank line at adin1110_read_fifo():310
